pa4/checker-pa4.ml

```diff
@@ -286,10 +307,21 @@ let add_method_types meth_env input_class = begin
 	List.iter(fun own_feat -> 
 		match own_feat with
 		|  Method ((_, m_name), formals, (_, return_type),_) -> 
-			let formal_types = List.map (fun (_, (_, fname)) -> Class fname) formals in
-			let key: string * string = (class_name, m_name) in
-			let value: static_type list= formal_types  @  [Class return_type] in
-			Hashtbl.add meth_env key value
+			let formal_types = List.map (fun (_,(_,fname)) -> 
+					if fname = class_name then 
+						SELF_TYPE fname
+					else 
+						Class fname
+				) formals in
+			let key = (class_name,m_name) in
+			let return_class = begin
+				if return_type = class_name then 
+					SELF_TYPE return_type
+				else 
+					Class return_type 
+			end in
+			let value = formal_types  @  [return_class] in
```

Comment:
Same thing that I said about above the key, I think that just including this on line 324 would be clearier.